### About the input file
#### This data set pertains to all types of structural permits from Jan 1, 2013-Feb 25th 2018. Data includes details on application/permit numbers, job addresses, supervisorial districts, and the current status of the applications. Data is uploaded weekly by DBI. Users can access permit information online through DBI’s Permit Tracking System which is 24/7 at www.sfdbi.org/dbipts.

In [4]:
import pandas as pd
from pandas import to_datetime

In [5]:
data = pd.read_csv('Building_Permits.csv')

/var/folders/m7/2d5d832j3j300wwm7q5ptzdh0000gn/T/ipykernel_12803/2139922528.py:1: DtypeWarning: Columns (22,32) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('Building_Permits.csv')


In [6]:
#Understand the data by looking at the first 10 rows
data.head(10)

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   
5  201706149344            8            otc alterations permit   
6  201706300814            8            otc alterations permit   
7       M803667            8            otc alterations permit   
8       M804227            8            otc alterations permit   
9       M804767            8            otc alterations permit   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           05/06/2015  0326  023            140                  NaN   
1           04/19/2016  0306  007            440                  NaN   
2           05/27/2016  0595  203           1647                  NaN   
3           11/07/2016  0156  011           1230                  NaN   
4           11/28/2016  0342  001            950                  NaN   
5           06/14/2017  4105  009            800                  NaN   
6           06/30/2017  1739  020           1291                  NaN   
7           06/30/2017  4789  014           1465                  NaN   
8           07/05/2017  1212  054           2094                  NaN   
9           07/06/2017  1259  016             89                  NaN   

  Street Name Street Suffix  ...  Existing Construction Type  \
0       Ellis            St  ...                         3.0   
1       Geary            St  ...                         3.0   
2     Pacific            Av  ...                         1.0   
3     Pacific            Av  ...                         5.0   
4      Market            St  ...                         3.0   
5     Indiana            St  ...                         1.0   
6        11th            Av  ...                         5.0   
7      Revere            Av  ...                         NaN   
8        Fell            St  ...                         NaN   
9      Alpine            Tr  ...                         NaN   

  Existing Construction Type Description Proposed Construction Type  \
0                          constr type 3                        NaN   
1                          constr type 3                        NaN   
2                          constr type 1                        1.0   
3                         wood frame (5)                        5.0   
4                          constr type 3                        NaN   
5                          constr type 1                        1.0   
6                         wood frame (5)                        5.0   
7                                    NaN                        NaN   
8                                    NaN                        NaN   
9                                    NaN                        NaN   

  Proposed Construction Type Description Site Permit Supervisor District  \
0                                    NaN         NaN                 3.0   
1                                    NaN         NaN                 3.0   
2                          constr type 1         NaN                 3.0   
3                         wood frame (5)         NaN                 3.0   
4                                    NaN         NaN                 6.0   
5                          constr type 1         NaN                10.0   
6                         wood frame (5)         NaN                 5.0   
7                                    NaN         NaN                10.0   
8                                    NaN         NaN                 5.0   
9                                    NaN         NaN                 8.0   

  Neighborhoods - Analysis Boundaries  Zipcode  \
0                          Tenderloin  94102.0   
1                          Tenderloin  94102.0 

In [7]:
data.dtypes

Permit Number                              object
Permit Type                                 int64
Permit Type Definition                     object
Permit Creation Date                       object
Block                                      object
Lot                                        object
Street Number                               int64
Street Number Suffix                       object
Street Name                                object
Street Suffix                              object
Unit                                      float64
Unit Suffix                                object
Description                                object
Current Status                             object
Current Status Date                        object
Filed Date                                 object
Issued Date                                object
Completed Date                             object
First Construction Document Date           object
Structural Notification                    object


In [8]:
#Find the number of rows and columns
data.shape

(198900, 43)

In [9]:
#Find missing values in the dataset
data.isnull().sum()

Permit Number                                  0
Permit Type                                    0
Permit Type Definition                         0
Permit Creation Date                           0
Block                                          0
Lot                                            0
Street Number                                  0
Street Number Suffix                      196684
Street Name                                    0
Street Suffix                               2768
Unit                                      169421
Unit Suffix                               196939
Description                                  290
Current Status                                 0
Current Status Date                            0
Filed Date                                     0
Issued Date                                14940
Completed Date                            101709
First Construction Document Date           14946
Structural Notification                   191978
Number of Existing S

In [10]:
#Find the percentage of missing data
for i in data.columns:
    percentage_missing_column = data[i].isnull().sum()/data.shape[0]
    print(f"Percentage of missing values in {i}: {round(percentage_missing_column*100, 2)}")

Percentage of missing values in Permit Number: 0.0
Percentage of missing values in Permit Type: 0.0
Percentage of missing values in Permit Type Definition: 0.0
Percentage of missing values in Permit Creation Date: 0.0
Percentage of missing values in Block: 0.0
Percentage of missing values in Lot: 0.0
Percentage of missing values in Street Number: 0.0
Percentage of missing values in Street Number Suffix: 98.89
Percentage of missing values in Street Name: 0.0
Percentage of missing values in Street Suffix: 1.39
Percentage of missing values in Unit: 85.18
Percentage of missing values in Unit Suffix: 99.01
Percentage of missing values in Description: 0.15
Percentage of missing values in Current Status: 0.0
Percentage of missing values in Current Status Date: 0.0
Percentage of missing values in Filed Date: 0.0
Percentage of missing values in Issued Date: 7.51
Percentage of missing values in Completed Date: 51.14
Percentage of missing values in First Construction Document Date: 7.51
Percentag

### Observation: Street number missing values in the dataset are 98.89%. You need to understand whether these are missing values because someone couldn't enter it or they don't even exist. What do you think?

### Street Number Suffix, Street Suffix, Unit, Unit Suffix: Not all addresses have these components.
### Structural Notification: Only required for certain legal requirements.
### Voluntary Soft-Story Retrofit: Only applicable if the building falls under retrofit regulations.
### Fire Only Permit: Not all permits are related to fire hazards.
### TIDF Compliance: Only applicable if the permit is subject to TIDF (Transit Impact Development Fee).
### Plansets: Some permits might not require plan submissions.
### Site Permit: Only applicable for large projects requiring a separate site approval.

In [11]:
data.columns

Index(['Permit Number', 'Permit Type', 'Permit Type Definition',
       'Permit Creation Date', 'Block', 'Lot', 'Street Number',
       'Street Number Suffix', 'Street Name', 'Street Suffix', 'Unit',
       'Unit Suffix', 'Description', 'Current Status', 'Current Status Date',
       'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Structural Notification',
       'Number of Existing Stories', 'Number of Proposed Stories',
       'Voluntary Soft-Story Retrofit', 'Fire Only Permit',
       'Permit Expiration Date', 'Estimated Cost', 'Revised Cost',
       'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',
       'Plansets', 'TIDF Compliance', 'Existing Construction Type',
       'Existing Construction Type Description', 'Proposed Construction Type',
       'Proposed Construction Type Description', 'Site Permit',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries', 'Zipcode',
       'Location', 'Record ID'],
     

In [12]:
#drop column TIDF Compliance since it is entirely missing
data.drop(['Street Number Suffix', 'Structural Notification', 'Site Permit', 'Unit', 'Unit Suffix', 'TIDF Compliance', 'Voluntary Soft-Story Retrofit', 'Fire Only Permit'], axis=1, inplace=True)

In [13]:
data.isnull().sum()

Permit Number                                  0
Permit Type                                    0
Permit Type Definition                         0
Permit Creation Date                           0
Block                                          0
Lot                                            0
Street Number                                  0
Street Name                                    0
Street Suffix                               2768
Description                                  290
Current Status                                 0
Current Status Date                            0
Filed Date                                     0
Issued Date                                14940
Completed Date                            101709
First Construction Document Date           14946
Number of Existing Stories                 42784
Number of Proposed Stories                 42868
Permit Expiration Date                     51880
Estimated Cost                             38066
Revised Cost        

In [14]:
data['Current Status'].unique()

array(['expired', 'issued', 'withdrawn', 'complete', 'filed', 'approved',
       'reinstated', 'cancelled', 'appeal', 'suspend', 'revoked',
       'plancheck', 'disapproved', 'incomplete'], dtype=object)

In [15]:
data.shape

(198900, 35)

In [16]:
data['Completed Date'].unique()

array([nan, '07/24/2017', '07/12/2017', ..., '11/02/2017', '12/26/2017',
       '11/01/2017'], dtype=object)

In [17]:
data['Completed Date'] = pd.to_datetime(data['Completed Date'])
data[(data['Current Status'] == 'complete') & (data['Completed Date'].isna())][['Current Status', 'Completed Date']]

Empty DataFrame
Columns: [Current Status, Completed Date]
Index: []

In [18]:
data[(data['Issued Date'].isna()) & (data['Completed Date'].isna())][['Current Status', 'Issued Date', 'Completed Date']]

Current Status Issued Date Completed Date
2           withdrawn         NaN            NaT
13              filed         NaN            NaT
18              filed         NaN            NaT
21              filed         NaN            NaT
22              filed         NaN            NaT
...               ...         ...            ...
198878          filed         NaN            NaT
198884          filed         NaN            NaT
198885          filed         NaN            NaT
198886       approved         NaN            NaT
198890          filed         NaN            NaT

[14921 rows x 3 columns]

In [19]:
data[~((data['Issued Date'].isna()) & (data['Completed Date'].isna()))]

Permit Number  Permit Type  Permit Type Definition  \
0       201505065519            4            sign - erect   
1       201604195146            4            sign - erect   
3       201611072166            8  otc alterations permit   
4       201611283529            6             demolitions   
5       201706149344            8  otc alterations permit   
...              ...          ...                     ...   
198895       M862628            8  otc alterations permit   
198896  201712055595            8  otc alterations permit   
198897       M863507            8  otc alterations permit   
198898       M863747            8  otc alterations permit   
198899       M864287            8  otc alterations permit   

       Permit Creation Date Block   Lot  Street Number Street Name  \
0                05/06/2015  0326   023            140       Ellis   
1                04/19/2016  0306   007            440       Geary   
3                11/07/2016  0156   011           1230     Pacific   
4                11/28/2016  0342   001            950      Market   
5                06/14/2017  4105   009            800     Indiana   
...                     ...   ...   ...            ...         ...   
198895           12/05/2017  0113  017A           1228  Montgomery   
198896           12/05/2017  0271   014            580        Bush   
198897           12/06/2017  4318   019           1568     Indiana   
198898           12/06/2017  0298   029            795      Sutter   
198899           12/07/2017  0160   006            838     Pacific   

       Street Suffix                                        Description  ...  \
0                 St  ground fl facade: to erect illuminated, electr...  ...   
1                 St            remove (e) awning and associated signs.  ...   
3                 Av           repair dryrot & stucco at front of bldg.  ...   
4                 St  demolish retail/office/commercial 3-story buil...  ...   
5                 St                                          evac maps  ...   
...              ...                                                ...  ...   
198895            St                                       street space  ...   
198896            St                fire alarm upgrade ref 201704123852  ...   
198897            St                                       street space  ...   
198898            St                                street space permit  ...   
198899            Av                                street space permit  ...   

       Plansets Existing Construction Type  \
0           2.0                        3.0   
1           2.0                        3.0   
3           2.0                        5.0   
4           2.0                        3.0   
5           2.0                        1.0   
...         ...                        ...   
198895      NaN                        NaN   
198896      2.0                        5.0   
198897      NaN                        NaN   
198898      NaN                        NaN   
198899      NaN                        NaN   

       Existing Construction Type Description Proposed Construction Type  \
0                               constr type 3                        NaN   
1                               constr type 3                        NaN   
3                              wood frame (5)                        5.0   
4                               constr type 3                        NaN   
5                               constr type 1                        1.0   
...                                       ...                        ...   
198895                                    NaN                        NaN   
198896                         wood frame (5)                        5.0   
198897                                    NaN                        NaN   
198898                                    NaN                        NaN   
198899                                    NaN                        NaN   

      

In [20]:
data.columns

Index(['Permit Number', 'Permit Type', 'Permit Type Definition',
       'Permit Creation Date', 'Block', 'Lot', 'Street Number', 'Street Name',
       'Street Suffix', 'Description', 'Current Status', 'Current Status Date',
       'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Number of Existing Stories',
       'Number of Proposed Stories', 'Permit Expiration Date',
       'Estimated Cost', 'Revised Cost', 'Existing Use', 'Existing Units',
       'Proposed Use', 'Proposed Units', 'Plansets',
       'Existing Construction Type', 'Existing Construction Type Description',
       'Proposed Construction Type', 'Proposed Construction Type Description',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries', 'Zipcode',
       'Location', 'Record ID'],
      dtype='object')

In [21]:
data.isnull().sum()

Permit Number                                  0
Permit Type                                    0
Permit Type Definition                         0
Permit Creation Date                           0
Block                                          0
Lot                                            0
Street Number                                  0
Street Name                                    0
Street Suffix                               2768
Description                                  290
Current Status                                 0
Current Status Date                            0
Filed Date                                     0
Issued Date                                14940
Completed Date                            101709
First Construction Document Date           14946
Number of Existing Stories                 42784
Number of Proposed Stories                 42868
Permit Expiration Date                     51880
Estimated Cost                             38066
Revised Cost        

In [22]:
data.dropna(subset=['Street Suffix','Plansets'], inplace=True)

### Mean and Median Imputation
#### Both mean and median imputation are techniques used to handle missing data by replacing missing values with a central tendency measure of the available data.

*1. Mean Imputation*
*Definition:* Missing values are replaced with the mean (average) of the non-missing values in the column.
*When to Use:*
When data is normally distributed (symmetrical).
When you want to preserve the mean of the dataset.
Works well for continuous numerical features.
*Limitations:*
Affected by outliers, which can skew the imputation.
Reduces variance in the dataset, leading to loss of variability.
*2. Median Imputation*
*Definition:* Missing values are replaced with the median (middle value) of the non-missing values.
*When to Use:*
When data is skewed or contains outliers.
Works well for continuous and ordinal data.
Helps in preserving the distribution of the data better than mean imputation.
*Limitations:*
Does not consider relationships with other variables.
Can still distort variability but less than mean imputation.

In [23]:
#Performing mean imputation on Estimated Cost
data['Estimated Cost'] = data['Estimated Cost'].fillna(data['Estimated Cost'].mean())

In [24]:
#Performing median imputation on Revised Cost
data['Revised Cost'] = data['Revised Cost'].fillna(data['Revised Cost'].median())

In [25]:
data.isnull().sum()

Permit Number                                 0
Permit Type                                   0
Permit Type Definition                        0
Permit Creation Date                          0
Block                                         0
Lot                                           0
Street Number                                 0
Street Name                                   0
Street Suffix                                 0
Description                                 189
Current Status                                0
Current Status Date                           0
Filed Date                                    0
Issued Date                               14394
Completed Date                            63509
First Construction Document Date          14406
Number of Existing Stories                 5412
Number of Proposed Stories                 5488
Permit Expiration Date                    14405
Estimated Cost                                0
Revised Cost                            

### Relationship between Status and Issued Date
#### If Status is Issued, Complete, Expired, Reinstated then Issued Date should be present.
#### If Status is Filed, Approved, Plancheck, Appeal then Issued Date may or maynot exist since Permit is still in review or not finalized yet.
#### If Status is Withdrawn, Cancelled, Revoked, Disapproved, Incomplete then Permit doesn't exist so it can be empty.
#### If Status is Suspend then Issue Date may or maynot exist. Since the Issue Date only exist if the permit has been issued previously.

In [26]:
print(data[data['Issued Date'].isna()]['Current Status'].value_counts())

Current Status
filed          11526
withdrawn       1740
approved         713
cancelled        337
reinstated        37
plancheck         16
complete          16
suspend            4
appeal             2
disapproved        2
incomplete         1
Name: count, dtype: int64


In [27]:
#Fill the missing values for 'Complete' or 'Expired' Status using the first construction document date.
data.loc[(data['Issued Date'].isna()) & (data['Current Status'] == 'complete'), 'Issued Date'] = data['First Construction Document Date']

In [28]:
data = data[~((data['Issued Date'].isna()) & (data['Current Status'].isin(['issued', 'complete', 'reinstated', 'expired'])))]

In [29]:
data.isnull().sum()

Permit Number                                 0
Permit Type                                   0
Permit Type Definition                        0
Permit Creation Date                          0
Block                                         0
Lot                                           0
Street Number                                 0
Street Name                                   0
Street Suffix                                 0
Description                                 189
Current Status                                0
Current Status Date                           0
Filed Date                                    0
Issued Date                               14341
Completed Date                            63472
First Construction Document Date          14354
Number of Existing Stories                 5403
Number of Proposed Stories                 5486
Permit Expiration Date                    14353
Estimated Cost                                0
Revised Cost                            

In [30]:
#Check the missing values in Number of existing stories and number of proposed stories based on the Current Status
print(data[data['Number of Existing Stories'].isna()]['Current Status'].value_counts())

Current Status
complete      1868
issued        1846
filed         1396
withdrawn      142
approved        54
cancelled       41
reinstated      29
expired         21
suspend          4
plancheck        1
revoked          1
Name: count, dtype: int64


In [31]:
print(data[data['Number of Proposed Stories'].isna()]['Current Status'].value_counts())


Current Status
issued         2121
complete       1629
filed          1367
cancelled       134
withdrawn       112
approved         73
expired          35
reinstated       13
disapproved       1
revoked           1
Name: count, dtype: int64


### Using ffill() and bfill()
Forward fill (ffill()) fills missing values using the previous available value in the same group.
In this case, This assumes that if a building's permit was previously recorded, the same structure details apply.
Backward fill (bfill()) fills missing values using the next available value in the same group.
This helps when the first row(s) of a group have missing values (since ffill() can't fill upwards).

### When Not to Use Both?
If missing values are random (not grouped by address), ffill() and bfill() may introduce errors.
If missing values are too common, consider median imputation instead.

In [32]:
data.columns

Index(['Permit Number', 'Permit Type', 'Permit Type Definition',
       'Permit Creation Date', 'Block', 'Lot', 'Street Number', 'Street Name',
       'Street Suffix', 'Description', 'Current Status', 'Current Status Date',
       'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Number of Existing Stories',
       'Number of Proposed Stories', 'Permit Expiration Date',
       'Estimated Cost', 'Revised Cost', 'Existing Use', 'Existing Units',
       'Proposed Use', 'Proposed Units', 'Plansets',
       'Existing Construction Type', 'Existing Construction Type Description',
       'Proposed Construction Type', 'Proposed Construction Type Description',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries', 'Zipcode',
       'Location', 'Record ID'],
      dtype='object')

In [33]:
data.sort_values(['Street Number', 'Street Name', 'Issued Date'], inplace=True)
data[['Number of Existing Stories', 'Number of Proposed Stories']] = data.groupby(['Street Number', 'Street Name'])[['Number of Existing Stories', 'Number of Proposed Stories']].ffill().bfill()

In [34]:
#### Fill missing values in Permit Expiration Date
print(data[['First Construction Document Date', 'Permit Expiration Date']].isna().sum())

First Construction Document Date    14354
Permit Expiration Date              14353
dtype: int64


In [35]:
#Check the current status for both the missing dates
print(data[data['First Construction Document Date'].isna()]['Current Status'].value_counts())

Current Status
filed          11515
withdrawn       1740
approved         707
cancelled        337
complete          23
plancheck         16
issued             7
suspend            4
appeal             2
disapproved        2
incomplete         1
Name: count, dtype: int64


In [36]:
print(data[data['Permit Expiration Date'].isna()]['Current Status'].value_counts())

Current Status
filed          11514
withdrawn       1740
approved         707
cancelled        337
plancheck         16
issued            16
complete          13
suspend            4
appeal             2
disapproved        2
reinstated         1
incomplete         1
Name: count, dtype: int64


In [37]:
#Handle the missing values with the median delay if the Current status is complete, issued or expired
data['First Construction Document Date'] = to_datetime(data['First Construction Document Date'])
data['Issued Date'] = to_datetime(data['Issued Date'])

# Compute median delay between issued date and first construction document date
median_delay = (data['First Construction Document Date'] - data['Issued Date']).median()

# Fill missing values for "issued" and "complete" permits
data.loc[(data['First Construction Document Date'].isna()) & (data['Current Status'].isin(['issued', 'complete', 'expired'])), 
         'First Construction Document Date'] = data['Issued Date'] + median_delay

In [38]:
#Filling the missing Permit Expiration date assuming a 1-year expiration date
data['Permit Expiration Date'] = to_datetime(data['Permit Expiration Date'])
data.loc[data['Permit Expiration Date'].isna(), 'Permit Expiration Date'] = data['Issued Date'] + pd.DateOffset(years=1)

In [39]:
data.dtypes

Permit Number                                     object
Permit Type                                        int64
Permit Type Definition                            object
Permit Creation Date                              object
Block                                             object
Lot                                               object
Street Number                                      int64
Street Name                                       object
Street Suffix                                     object
Description                                       object
Current Status                                    object
Current Status Date                               object
Filed Date                                        object
Issued Date                               datetime64[ns]
Completed Date                            datetime64[ns]
First Construction Document Date          datetime64[ns]
Number of Existing Stories                       float64
Number of Proposed Stories     

In [40]:
data.isnull().sum()

Permit Number                                 0
Permit Type                                   0
Permit Type Definition                        0
Permit Creation Date                          0
Block                                         0
Lot                                           0
Street Number                                 0
Street Name                                   0
Street Suffix                                 0
Description                                 189
Current Status                                0
Current Status Date                           0
Filed Date                                    0
Issued Date                               14341
Completed Date                            63472
First Construction Document Date          14324
Number of Existing Stories                    0
Number of Proposed Stories                    0
Permit Expiration Date                    14323
Estimated Cost                                0
Revised Cost                            

### Check missing values in existing use:
Missing values are acceptable if Permit is for empty land or a new development.
Missing values are not acceptable if the permit is for modifications, but no existing use is listed.

In [41]:
#If the same street number/name has multiple permits, we can assume the existing/proposed use and units should be similar.
data[['Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units']] = (
    data.groupby(['Street Number', 'Street Name'])[['Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units']]
    .ffill()
    .bfill()
)


### Check missing values in existing units:
Missing values are acceptable if Permit applies to a single-use commercial/industrial property (no units).
Missing values are not acceptable if the property is residential, but units are missing.

In [42]:
# Assume default values based on building type
default_units = {
    'Single Family': 1,
    'Multi-Family': 2,  # Assume minimum of 2 units
    'Apartment': 10,  # Average for small apartment buildings
    'Commercial': 0,
    'Industrial': 0
}

# Fill missing existing units based on existing use
data['Existing Units'] = data.apply(
    lambda row: default_units.get(row['Existing Use'], np.nan) if pd.isna(row['Existing Units']) else row['Existing Units'],
    axis=1
)

### Check missing values in proposed use:
Missing values are acceptable if Permit was withdrawn/cancelled, so no proposed use exists.
Missing values are not acceptable if the permit was approved/issued, but no proposed use is recorded.

In [43]:
#Use mode for the area to fill in the missing values
data['Proposed Use'] = data.groupby('Zipcode')['Proposed Use'].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else "Unknown"))

### Check missing values in proposed units:
Missing values are acceptable if Permit is for repairs or minor modifications that don’t change unit count.
Missing values are not acceptable if the permit Permit is for repairs or minor modifications that don’t change unit count.

In [44]:
data = data.dropna(subset=['Proposed Units'])

In [45]:
#Remove all those rows where location is not null but zipcode is null
data = data[~((data['Location'].notnull()) & (data['Zipcode'].isnull()))]

In [46]:
data[['Latitude', 'Longitude']] = data['Location'].str.extract(r'\(([^,]+), ([^)]+)\)').astype(float)

In [47]:
data['Latitude']

51215           NaN
106588    37.787079
149257    37.787079
181918    37.786807
22862     37.787056
            ...    
65704     37.778683
107325    37.778679
118418    37.710482
55716     37.710482
102439    37.710482
Name: Latitude, Length: 159243, dtype: float64

In [48]:
print(data[['Latitude', 'Longitude']].isna().sum())

Latitude     1079
Longitude    1079
dtype: int64


In [49]:
#Handling missing values in Location using the Geopy
from geopy.geocoders import Nominatim
import time

geolocator = Nominatim(user_agent="sf_permit_locator")

def geocode_location(row):
    if pd.isna(row['Latitude']) or pd.isna(row['Longitude']):
        address = f"{row['Street Number']} {row['Street Name']}, San Francisco, CA {row['Zipcode']}"
        try:
            location = geolocator.geocode(address, timeout=10)
            if location:
                return pd.Series([location.latitude, location.longitude])
        except Exception as e:
            print(f"Geocoding failed for {address}: {e}")
            return pd.Series([None, None])
    return pd.Series([row['Latitude'], row['Longitude']])

ModuleNotFoundError: No module named 'geopy'

In [ ]:
data[['Latitude', 'Longitude']] = data.apply(geocode_location, axis=1)

KeyboardInterrupt: 